src/serializer/ResidualHeapSerializer.js

```diff
@@ -0,0 +1,1401 @@
+      this.preludeGenerator.createNameGenerator("__scope_"),
+      residualFunctionInfos
+    );
+    this.residualHeapInspector = residualHeapInspector;
+    this.residualValues = residualValues;
+    this.residualFunctionBindings = residualFunctionBindings;
+    this.residualFunctionInfos = residualFunctionInfos;
+
+    this.collectValToRefCountOnly = collectValToRefCountOnly;
+    this.valToRefCount = valToRefCount;
+  }
+
+  globalReasons: {
+    [filename: string]: Array<string>,
+  };
+
+  declarativeEnvironmentRecordsBindings: Map<VisitedBinding, SerializedBinding>;
+  serializationStack: Array<Value>;
+  delayedSerializations: Array<{ body: Array<BabelNodeStatement>, func: () => void }>;
+  delayedKeyedSerializations: Map<
+    BabelNodeIdentifier,
+    Array<{ body: Array<BabelNodeStatement>, values: Array<Value>, func: () => void }>
+  >;
+  functions: Map<BabelNodeBlockStatement, Array<FunctionInstance>>;
+  functionInstances: Array<FunctionInstance>;
+  //value to intermediate references generated like $0, $1, $2,...
+  refs: Map<Value, BabelNodeIdentifier>;
+  collectValToRefCountOnly: boolean;
+  valToRefCount: void | Map<Value, number>;
+  prelude: Array<BabelNodeStatement>;
+  body: Array<BabelNodeStatement>;
+  mainBody: Array<BabelNodeStatement>;
+  realm: Realm;
+  declaredDerivedIds: Set<BabelNodeIdentifier>;
+  preludeGenerator: PreludeGenerator;
+  generator: Generator;
+  descriptors: Map<string, BabelNodeIdentifier>;
+  needsEmptyVar: boolean;
+  needsAuxiliaryConstructor: boolean;
+  valueNameGenerator: NameGenerator;
+  descriptorNameGenerator: NameGenerator;
+  factoryNameGenerator: NameGenerator;
+  logger: Logger;
+  modules: Modules;
+  requireReturns: Map<number | string, BabelNodeExpression>;
+  statistics: SerializerStatistics;
+  timingStats: TimingStatistics;
+  residualHeapInspector: ResidualHeapInspector;
+  residualValues: Map<Value, Set<Scope>>;
+  residualFunctionBindings: Map<FunctionValue, VisitedBindings>;
+  residualFunctionInfos: Map<BabelNodeBlockStatement, FunctionInfo>;
+  serializedValues: Set<Value>;
+  residualFunctions: ResidualFunctions;
+
+  _getBodyReference() {
+    return new BodyReference(this.body, this.body.length);
+  }
+
+  addProperties(
+    name: string,
+    obj: ObjectValue,
+    reasons: Array<string>,
+    alternateProperties: ?Map<string, PropertyBinding>,
+    objectPrototypeAlreadyEstablished: boolean = false
+  ) {
+    /*
+    for (let symbol of obj.symbols.keys()) {
+      // TODO #22: serialize symbols
+    }
+    */
+
+    // inject properties
+    for (let [key, propertyBinding] of alternateProperties || obj.properties) {
+      invariant(propertyBinding);
+      let desc = propertyBinding.descriptor;
+      if (desc === undefined) continue; //deleted
+      if (this.residualHeapInspector.canIgnoreProperty(obj, key)) continue;
+      invariant(desc !== undefined);
+      this._eagerOrDelay(this._getDescriptorValues(desc).concat(obj), () => {
```

Comment:
_getDescriptorValues should probably be more consistent with _getNestedAbstractValues and take [obj] as a second parameter. concat is not great for performance here.